src/ucp/wireup/wireup_cm.c

```diff
@@ -0,0 +1,272 @@
+    if (status != UCS_OK) {
+        goto free_ucp_addr;
+    }
+
+    addr_indices = ucs_malloc(unpacked_addr.address_count *
+                              sizeof(*addr_indices), "dummy addr indices");
+    if (addr_indices == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto free_addr_list;
+    }
+
+    status = ucp_wireup_select_lanes(ucp_ep, &wireup_ep->ucp_ep_params,
+                                     ucp_cm_ep_init_flags(worker,
+                                                          &wireup_ep->ucp_ep_params),
+                                     unpacked_addr.address_count,
+                                     unpacked_addr.address_list, addr_indices,
+                                     &key);
+    if (status != UCS_OK) {
+        goto free_addr_indicies;
+    }
+
+    status = ucp_worker_get_ep_config(worker, &key, 0, &ucp_ep->cfg_index);
+    if (status != UCS_OK) {
+        goto free_addr_indicies;
+    }
+
+    ucp_ep->am_lane = key.am_lane;
+
+    /* 0-lane is a CM lane */
+    for (lane_idx = 1; lane_idx < ucp_ep_num_lanes(ucp_ep); ++lane_idx) {
+        status = ucp_wireup_ep_create(ucp_ep, &ucp_ep->uct_eps[lane_idx]);
+        if (status != UCS_OK) {
+            break;
+        }
+    }
+
+    if (status != UCS_OK) {
+        for (lane_idx = 1; (lane_idx < ucp_ep_num_lanes(ucp_ep)) &&
+                           (ucp_ep->uct_eps[lane_idx] != NULL); ++lane_idx) {
+            uct_ep_destroy(ucp_ep->uct_eps[lane_idx]);
+        }
+    }
+
+free_addr_indicies:
+    ucs_free(addr_indices);
+free_addr_list:
+    ucs_free(unpacked_addr.address_list);
+free_ucp_addr:
+    ucs_free(ucp_addr);
+out:
+    return status;
+}
+
+static ssize_t ucp_cm_client_priv_pack_cb(void *arg, const char *dev_name,
+                                          void *priv_data)
+{
+    ucp_wireup_sockaddr_data_t *sa_data = priv_data;
+    ucp_ep_h ep                         = arg;
+    ucp_worker_h worker                 = ep->worker;
+    uct_cm_h cm                         = worker->cms[/*cm_idx = */ 0].cm;
+    uint64_t tl_bitmap;
+    ucp_wireup_ep_t *wireup_ep;
+    uct_ep_h tl_ep;
+    uct_cm_attr_t cm_attr;
+    uct_ep_params_t tl_ep_params;
+    void* ucp_addr;
+    size_t ucp_addr_size;
+    ucs_status_t status;
+    ucp_lane_index_t lane_idx;
+    ucp_rsc_index_t rsc_idx;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    wireup_ep = ucp_ep_get_cm_wireup_ep(ep);
+    ucs_assert_always(wireup_ep != NULL);
+    strncpy(wireup_ep->dev_name, dev_name, UCT_DEVICE_NAME_MAX);
+    wireup_ep->dev_name[UCT_DEVICE_NAME_MAX - 1] = '\0';
+
+    status = ucp_cm_preinit_client_lanes(ep);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    cm_attr.field_mask = UCT_CM_ATTR_FIELD_MAX_CONN_PRIV;
+    status             = uct_cm_query(cm, &cm_attr);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    tl_bitmap = 0;
+    for (lane_idx = 0; lane_idx < ucp_ep_num_lanes(ep); ++lane_idx) {
```

Comment:
seems a bit weird we are going over the lanes twice. probably need to unite this code with lines 97-102 of ucp_cm_preinit_client_lanes

Reply:
fixed in #4283